### ISSUE TITLE
CMake find_package finds system lib instead of Conan-packaged library on Android and iOS

### ISSUE THREAD

**Issue by ssrobins** [conan-io/conan#5437]
Environment:
Conan 1.16.1
CMake 3.15.0-rc3
Android NDK r20
Xcode 10.2.1
macOS 10.14.4

~/.bash_profile contains:
export ANDROID_HOME=/Users/steve/Code/android
export android_ndk_version=r20

Android NDK installed at:
$ANDROID_HOME/android-ndk-r20/

Reproduction steps:

git clone https://gitlab.com/ssrobins/conan-libpng.git
cd conan-libpng

Run to build package on iOS:
./build_ios.sh

Run to build package on Android:
./build_android-v8a.sh

On Android, CMake finds zlib from the Android NDK toolchain: /root/android-ndk-r20/toolchains/llvm/prebuilt/linux-x86_64/sysroot/usr/lib/arm-linux-androideabi/libz.a

On iOS, it finds zlib from Xcode:
/Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS12.2.sdk/usr/lib/libz.tbd

When running the build scripts for Windows, Linux, and Mac, it does find the optimized and debug libs from my Conan zlib package, as expected.

I would expect it to find the zlib package I reference in the requirements method in my conanfile.py on all platforms, not just the desktop platforms.

Full output of each build showing the issues:
https://gitlab.com/ssrobins/conan-libpng/pipelines/69101595

In my top-level CMakeLists.txt file (https://gitlab.com/ssrobins/conan-libpng/blob/stable/CMakeLists.txt), I can add this:
`set(CMAKE_FIND_ROOT_PATH ${CONAN_ZLIB_ROOT} ${CMAKE_FIND_ROOT_PATH})`

Then the Conan packaged zlib is found on all platforms:
https://gitlab.com/ssrobins/conan-libpng/pipelines/69106660

I see the conan_set_find_paths macro in conanbuildinfo.cmake but it requires me to set variables for it to work, which is similar to my workaround above.  The conanbuildinfo.cmake contains all the info to just take care of this for me automatically so it would be nice if it did that.

**Comment by ssrobins** [comment]
I can use this line instead so I still only have to set one variable even if there are multiple dependencies:
`set(CMAKE_FIND_ROOT_PATH ${CONAN_CMAKE_MODULE_PATH} ${CMAKE_FIND_ROOT_PATH})`

**Comment by danimtb** [comment]
Hi @ssrobins and thanks for the report.
I agree the situation with the different possibilities of CMake generators and the CMake helper is a bit confusing and we are working to unify them.

I was checking the issue and according to the code and the [documentation](https://docs.conan.io/en/latest/reference/generators/cmake.html#conan-set-find-paths) at the CMake macro `conan_set_find_paths()` should be applied if you are using the `conan_basic_setup()` on you *CMakeLists.txt*, do you why is that not working?

Alternatively, as you said, you could set the `set(CMAKE_FIND_ROOT_PATH ${CONAN_CMAKE_MODULE_PATH} ${CMAKE_FIND_ROOT_PATH})` or add the *cmake_paths* to the list of generators in your recipe https://docs.conan.io/en/latest/integrations/build_system/cmake/cmake_paths_generator.html

**Comment by ssrobins** [comment]
In the generated `conanbuildinfo.cmake`, I see that the following command is set in the `conan_set_find_paths()` macro, which is run by `conan_basic_setup()`:
```
set(CMAKE_FIND_ROOT_PATH ${CONAN_CMAKE_FIND_ROOT_PATH} ${CMAKE_FIND_ROOT_PATH})
```

When I print out the value of `${CONAN_CMAKE_FIND_ROOT_PATH}` after `conan_basic_setup()` is called, it's blank.  Should the variable be `${CONAN_CMAKE_MODULE_PATH}` instead?

When I tried adding `cmake_paths` to the list of generators in `conanfile.py` and commented out my workaround, it didn't fix the problem.

Keep in mind that this worked fine even with just the `cmake` generator on desktop platforms, but didn't work on mobile (cross build) platforms, in case that matters.

**Comment by ssrobins** [comment]
The workaround didn't work on freetype where find_package is used to find bzip2.  The system lib is found instead of the Conan packaged lib on iOS, Linux, and Mac.

Looks like the community version of the freetype Conan package process is finding the bzip Conan packaged lib in the mac build:
https://travis-ci.com/bincrafters/conan-freetype/jobs/210367925
But who knows if the travis-ci environment is customized to do that.  In my very vanilla macOS+Xcode environment, it doesn't work.

If you have any clues, I'd love to know.

**Comment by SSE4** [comment]
@ssrobins could you check your CMake toolchain file?
especially, if it contains these:
```
CMAKE_FIND_ROOT_PATH_MODE_INCLUDE
CMAKE_FIND_ROOT_PATH_MODE_LIBRARY
CMAKE_FIND_ROOT_PATH_MODE_PACKAGE
CMAKE_FIND_ROOT_PATH_MODE_PROGRAM
```
they need to be set to `BOTH` in order to find conan libraries, which aren't inside the sysroot.

**Comment by v-strob** [comment]
For iOS, I use the new built-in toolchain so I have no idea where to find that.  For Android, I use the toolchain that comes with Android NDK r20.  It sets INCLUDE, LIBRARY, and PACKAGE to ONLY and PROGRAM to NEVER.  On both iOS and Android, if I set the following at the top of my CMakeLists.txt file:

```
CMAKE_FIND_ROOT_PATH_MODE_INCLUDE
CMAKE_FIND_ROOT_PATH_MODE_LIBRARY
CMAKE_FIND_ROOT_PATH_MODE_PACKAGE
CMAKE_FIND_ROOT_PATH_MODE_PROGRAM
```

Nothing happens.

As an update to my last post, I found out why find_package wasn't working on bzip2.  The FindBZip2.cmake included in the CMake install specifies `bz2 bzip2` as the valid library names.  It will cycle through all the paths set in `CMAKE_FIND_ROOT_PATH` with the first name variation, bz2 and if it doesn't find it, it'll look in the second path, which is the sysroot path, where it may find a non-Conan lib.  If I name the library bz2, the lib is found in the Conan path.  This is an unfortunate default behavior of find_library in CMake.

So in summary, if you were to include this line in the conanbuildinfo.cmake provided for the cmake generator:
```
set(CMAKE_PREFIX_PATH ${CONAN_CMAKE_MODULE_PATH} ${CMAKE_PREFIX_PATH})
```

Then find_package will find all my Conan libs, as expected.  I think this will help Conan 'just work' in situations where the library your are trying to package uses find_package to get its dependencies.

**Comment by ssrobins** [comment]
[Recent changes](https://github.com/conan-io/cmake-conan/blob/9baed3a9b18c04fec75601d047e10b0a00432408/conan_provider.cmake#L278) to [CMake wrapper for Conan](https://github.com/conan-io/cmake-conan/tree/develop2) sets `CMAKE_FIND_ROOT_PATH_BOTH` in the `find_package` call inside `conan_provider.cmake` so the user does not need any special settings in their CMake code for `find_package` to find Conan packages. Closing.
